Add `filter` option

kmalakoff opened the pull request:
Added pattern matching as discussed in https://github.com/sindresorhus/p-event/issues/5.

I added an overload for filter function or option since the referred examples and my use cases did not need any of the other options.

I ensured it worked with multiArgs and error cases, but am not sure if there needs to be more combinations tested. Let me know!

---

Fixes #5

sindresorhus commented:
Great!

It should be an option instead of overriding the options argument.

Also needs to be documented.

sindresorhus commented on index.js:
```diff
@@ -1,10 +1,15 @@
 	let cancel;
 
 	const ret = new Promise((resolve, reject) => {
+		if (isFunction(opts)) {
```
No need for Lodash, just `if (typeof opts === 'function')`.

sindresorhus commented on index.js:
```diff
@@ -21,6 +26,16 @@ module.exports = (emitter, event, opts) => {
 		removeListener = removeListener.bind(emitter);
 
 		const resolveHandler = function (value) {
```
This whole function can be simplified to:

```js
const resolveHandler = function (value) {
	if (opts.multiArgs) {
		value = [].slice.apply(arguments);
	}

	if (opts.filter && !opts.filter(value)) {
		return;
	}

	cancel();
	resolve(value);
};
```

kmalakoff commented:
Not sure if I agree on the options-only approach. Here's what I'm thinking...the example that I was pointed to from the original issue and the issue I surfaced this feature in just had a filter function without options which I think is probably the common use case (eg. from previously writing a utility function for this, I believe most of the time there will be no other options used with the filter function). 

I prefer the API to be streamlined for the expected use case (eg. just like in many lodash functions, there are rarely options to simple utilities). Of course, it may just be my personal preference. Agreed or is options-only a required change to this pull request?

kmalakoff commented:
Updated with:

1) Replaced lodash.isfunction with `if (typeof opts === 'function')`
2) Simplified conditionals
3) Added documentation in readme.md

On the documentation, I only documented the happy path rather than all combinations since the combinations should be clear from the tests.

sindresorhus commented:
> Simplified conditionals

Look at my code block in https://github.com/sindresorhus/p-event/pull/6/files#r120651565 again. Your code can be simplified further.

sindresorhus commented:
I don't feel strongly about only having it as an option, so I'm ok with having this shortcut, but it needs to be better documented. Maybe adding `### pEvent(emitter, event, filter)` on line 46.

sindresorhus commented on readme.md:
```diff
@@ -101,6 +101,23 @@ Default: `Infinity`
+const pEvent = require('p-event');
+const emitter = require('./some-event-emitter');
+
+pEvent(emitter, '🦄', e => e >= 3).then(result => {
```
`e` => `event`. I prefer being verbose in examples.

kmalakoff commented:
@sindresorhus thank you for the feedback. 

Unfortunately, I've started a new job recently and have been having trouble getting to this. The changes you are suggesting seem pretty minor and arbitrarily opinionated so feel free to make them rather than waiting for me. Thank you!

kmalakoff commented:
I should say that I will eventually get to this, but it has just been a very busy period for a number of reasons and every time I've planned to look at this, something else has come up.

sindresorhus commented:
Sure, no worries. I can do the changes.

kmalakoff commented:
Awesome! Thank you so much (and I'm sorry for the delays).
